### PR TITLE
S3 Directory provider

### PR DESCRIPTION
**Pull request by bielu**
As I actually needed to move my indexes on v8 to s3 I started work on s3 directory provider for Examine. :)

### CODE REVIEW

**Comment by bielu** [comment]
Hi @Shazwazza I just pushed working for my version, if you have any comments please leave and I will fix them :)

File: src/Examine.S3Directory/AzureDirectoryNativeLockFactory.cs [hunk 1]
```diff
@@ -0,0 +1,47 @@
+//using System.IO;
+//using Lucene.Net.Store;
+
+//namespace Examine.AzureDirectory
```
Comment (Shazwazza) [inline src/Examine.S3Directory/AzureDirectoryNativeLockFactory.cs]: Do we need thsi commented out class?

File: src/Examine.S3Directory/AzureNativeLock.cs [hunk 2]
```diff
@@ -0,0 +1,446 @@
+﻿//using System;
+//using System.Diagnostics;
+//using System.IO;
+//using System.Threading;
+//using Lucene.Net.Store;
+//using Microsoft.WindowsAzure.Storage;
+//using Microsoft.WindowsAzure.Storage.Blob;
+
+//namespace Examine.AzureDirectory
```
Comment (Shazwazza) [inline src/Examine.S3Directory/AzureNativeLock.cs]: as above, is this commented out class needed?

File: src/Examine.S3Directory/S3Directory.cs [hunk 3]
```diff
@@ -0,0 +1,450 @@
...
+        private bool _inSync;
+        private readonly object _locker = new object();
+
+        protected internal readonly string _containerName;
+        private string _bucketURL;
+        protected internal AmazonS3Client _blobClient;
+        private readonly LockFactory _lockFactory;
+
+        /// <summary>
+        /// Create an AzureDirectory
```
Comment (Shazwazza) [inline src/Examine.S3Directory/S3Directory.cs]: comments don't really match what is actually being implemented :)

File: src/Examine.S3Directory/S3Directory.cs [hunk 4]
```diff
@@ -0,0 +1,450 @@
...
+using Amazon.S3.Util;
+using Examine.LuceneEngine.Directories;
+using Lucene.Net.Index;
+using Lucene.Net.Store;
+using Directory = Lucene.Net.Store.Directory;
+
+namespace Examine.S3Directory
+{
+    /// <summary>
+    /// A Lucene directory used to store master index files in blob storage and sync local files to a %temp% fast drive storage
```
Comment (Shazwazza) [inline src/Examine.S3Directory/S3Directory.cs]: blog storage?

**Comment by bielu** [comment]
@Shazwazza I will go through comments today and update and explain more about s3 API :)

**Comment by bielu** [comment]
@Shazwazza I updated that to provide more direct information about we are using s3, I corrected comments and replace name blob with s3Object as AWS calling blobs "Objects".
Implementation is basically a replacement of all azure APIs with AWS s3 API, exactly like you noticed. I will provide testing cases for you soon.

**Comment by Shazwazza** [comment]
asesome sounds great

**Comment by bielu** [comment]
@Shazwazza Actually here is few ways to test that without having s3 setup in aws:
The easiest one is: 
https://github.com/localstack/localstack
Here are even more options:
https://stackoverflow.com/questions/23991694/aws-s3-local-server-for-integration-testing
but I would recommend using LocalStack.
Behaviour is the same as in Blob storage. 
I was testing that using with Demo website :)

**Comment by Shazwazza** [comment]
Cool, hopefully i'll find some time to test soon ish. 

In the meantime, do you think it's worth just releasing the Azure blob storage provider as-is instead of just listing this purely as a proof of concept? I figure i should just release it and then people can start using it and if there's bugs they'll lodge them.

**Comment by Shazwazza** [comment]
I published it ;) https://github.com/Shazwazza/Examine/releases/tag/v0.1.90-azure-directory

**Comment by bielu** [comment]
as agreed closing as directories becaming  separate repo :)

**Pull request closed without merging**